### PR TITLE
Migrating to flat hash map [left and all]

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?
Last PR of replacing std::unordered_map with absl::flat_hash_map.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
For the core worker interface part (requiring flat hash map instead of unordered map), would there be any implication regarding cpp frontend APIs? cc @scv119 @SongGuyang

**Comment by SongGuyang** [comment]
> For the core worker interface part (requiring flat hash map instead of unordered map), would there be any implication regarding cpp frontend APIs? cc @scv119 @SongGuyang

Thank you for this indication. I think this change will break the compatibility of C++ API. Can we keep to use C++ standard library, instead of absl?

**Comment by WangTaoTheTonic** [comment]
@SongGuyang @rkooo567 For the api part, we can keep using unordered_map, and migrate everything else to flat_hash_map?

**Comment by rkooo567** [comment]
I'd imagine it'd be ideal to keep all core worker interfaces to language workers to use an unordered map? @SongGuyang does it sound reasonable?

**Comment by SongGuyang** [comment]
Yep, I think the basic principle is keeping the API. But we should also consider the performance. If we use `absl::flat_hash_map` in the Core Worker, we should convert `std::unordered_map` to `absl::flat_hash_map`  anyway?

**Comment by rkooo567** [comment]
Yeah I believe the PR is doing that now (it is done from Cython layer). Maybe we can either not use unordered map on core worker or do the conversion only for the interface. cc @scv119 what's your thought here?

File: python/ray/includes/absl.pxd [hunk 1]
```diff
@@ -0,0 +1,74 @@
+from libcpp.utility cimport pair
+
+cdef extern from "absl/container/flat_hash_map.h" namespace "absl" nogil:
```
Comment (scv119) [inline python/ray/includes/absl.pxd:3]: How do we ensure this is correct?
seems to me a bunch of functions are missing here, such as `merge` and `contains`

**Comment by scv119** [comment]
I'm leaning towards keeping CoreWorker interface (and any interface between C++ and python) using std libs; unless some cpython expert can confirm we are introducing absl::* into cython the right way.

**Comment by scv119** [comment]
Sorry need to be a bit cautious here. But thanks for clean up the code base!

**Pull request closed without merging**